coin select: add coin selection strategy option to all on-chain RPCs

mohamedawnallah opened the pull request:
## Change Description
This PR adds coin selection strategy option to the following on-chain RPCs `FundPsbt`, `SendOutputs`, `BatchOpenChannel`, `SendCoins`, `SendMany`, `EstimateFee` and add `coin_selection_strategy` optional parameter to `lncli` equivalent commands (if any).

Fixes #8498

## Steps to Test
Steps for reviewers to follow to test the change.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

Chinwendu20 commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -845,13 +853,24 @@ func batchOpenChannel(ctx *cli.Context) error {
+	// these are currently the supported strategies.
+	coinSelectionStrategy := ctx.String("coin_selection_strategy")
+	if coinSelectionStrategy != "largest" &&
+		coinSelectionStrategy != "random" {
```
I am wondering if a switch statement would be better here, that way as coinSelection strategy grows we can just add it as a new case?

Chinwendu20 commented on rpcserver.go:
```diff
@@ -1068,7 +1069,8 @@ func allowCORS(handler http.Handler, origins []string) http.Handler {
+	strategy wallet.CoinSelectionStrategy) (*chainhash.Hash, error) {
 
 	outputs, err := addrPairsToOutputs(paymentMap, r.cfg.ActiveNetParams.Params)
 	if err != nil {
```
I think you would want to add the strategy to the dry run here as well, https://github.com/lightningnetwork/lnd/blob/ea2d5373fa4a0d76c6990302ca5d5f864d858c88/rpcserver.go#L1082

Chinwendu20 reviewed (commented):
Nice PR, needs release doc. I think your changes to the protobuf go files can be in the same commit where you made corresponding change to the proto file. Might help fix lint.

guggero commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -817,6 +818,13 @@ var batchOpenChannelCommand = cli.Command{
 		},
+		cli.StringFlag{
+			Name:  "coin_selection_strategy",
+			Usage: "(optional) The strategy to use for selecting " +
```
This should mention that when using the parameter, the globally (in `lnd.conf`) configured strategy is overwritten. I think that's one more reason to use an enum on the RPC side (see my other comments).

guggero commented on cmd/lncli/commands.go:
```diff
@@ -229,12 +237,23 @@ func estimateFees(ctx *cli.Context) error {
 	}
 
+	// Parse coin selection strategy.
+	coinSelectionStrategy := ctx.String("coin_selection_strategy")
```
Can extract this into a common parsing function that returns the enum value.

guggero commented on lnrpc/lightning.proto:
```diff
@@ -1105,6 +1105,11 @@ message EstimateFeeRequest {
+    // CoinSelectionStrategy is the strategy to use for selecting coins
+    // to send bitcoin on-chain to multiple addresses for getting the fee
+    // estimates.
+    string coin_selection_strategy = 5;
```
This should not be a string on the RPC level but an enum.
Then we can have the following values:

```
enum CoinSelectionStrategy {
    STRATEGY_USE_GLOBAL_CONFIG = 0; // Use whatever is configured in lnd.conf.
    STRATEGY_LARGEST = 1;
    STRATEGY_RANDOM = 2;
}
```

Then on the CLI we can have a string value that is parsed into the RPC enum values. The string values can then be `"global-config", "largest", "random"` with `"global-config"` being the default.

guggero replied:
This new enum should be defined in `lightning.proto` and re-used in all other proto files (by importing `lightning.proto`, if not already done).

guggero commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -811,12 +811,26 @@ func (w *WalletKit) SendOutputs(ctx context.Context,
 	}
 
+	// Parse coin selection strategy.
+	var coinSelectionStrategy base.CoinSelectionStrategy
```
We'll need a general parsing function that parses the new enum into the correct type. Maybe we can pass in the globally configured value and it will then decide which one to use.
A good place for that function would be `lnrpc/marshall_utils.go`.

guggero reviewed (commented):
Thanks for the PR! There's a lot of duplicated code in the CLI, see my inline comments on how that can be optimized.

mohamedawnallah commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,57 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+	switch strategy {
+	case CoinSelectionStrategy_STRATEGY_USE_GLOBAL_CONFIG:
+		strategyValue, _ := ParseCoinSelectionStrategyString(
+			"largest",
```
@guggero Thanks a lot for your feedback the code looks much cleaner because of the usage of common parsing functions. Now something there is one thing left here in `lnrpc/marshall_utils.go` when I use `lnd.DefaultConfig().CoinSelectionStrategy` instead of `largest` it gives me Import Cycle dependency issue because I needed to import `github.com/lightningnetwork/lnd` at the top of the file. I'm not used much to these issues any suggestions would be helpful, Thanks 🙏

guggero replied:
That's why I suggested to pass in the configured strategy as a parameter. Then the caller will find out what it is (`lnd.DefaultConfig().CoinSelectionStrategy` is wrong anyway, because that's not what the user configured but the default value if nothing is set by the user).

guggero commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,57 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+		)
+
+	case CoinSelectionStrategy_STRATEGY_LARGEST:
+		coinSelectionStrategy = wallet.CoinSelectionLargest
```
nit: can return directly in the `case` statements, then no final `return` at the end of the function is needed (since there's a `default` case).

guggero commented on funding/batch_test.go:
```diff
@@ -353,10 +353,12 @@ func TestBatchFund(t *testing.T) {
+				Channels:              tc.channels,
+				SatPerVbyte:           5,
+				MinConfs:              1,
+				CoinSelectionStrategy: strategy,
```
Because the default behavior now is to fall back to the pre-configured value, we don't need to update all those itests, wich makes the diff even smaller.

mohamedawnallah replied:
Thanks this makes total sense since I'm new to protobuf I found we could use the first value in the enum as the default one which in our case would be `STRATEGY_USE_GLOBAL_CONFIG`.

guggero commented on lnrpc/lightning.proto:
```diff
@@ -1091,6 +1091,12 @@ message LightningAddress {
 }
 
+enum CoinSelectionStrategy {
+    STRATEGY_USE_GLOBAL_CONFIG = 0;  // Use whatever is configured in lnd.conf.
```
nit: needs proper comments for each value, see other enums.

guggero commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,57 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+
+// Parse the coin selection strategy string from the command line
+// to an equivalent lnrpc.CoinSelectionStrategy enum type.
+func ParseCoinSelectionStrategyString(strategy string)(
```
nit: please check our formatting guidelines: https://github.com/lightningnetwork/lnd/blob/master/docs/code_formatting_rules.md

guggero replied:
Also, the "parse" is usually used to parse strings. So the function name doesn't need to have the word "string" in it.

mohamedawnallah replied:
Thanks, I missed this formatting guideline comment:
```
Function comments must always begin with the name of the function per [Effective Go](http://golang.org/doc/effective_go.html)
```

guggero commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,57 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+
+// Parse the coins selection strategy lnrpc.CoinSelectionStrategy enum type
+// to an equivalent wallet.CoinSelectionStrategy type.
+func ParseCoinSelectionStrategyEnum(strategy CoinSelectionStrategy)(
```
This kind of parsing is what we usually call "Unmarshaling" in the context of converting between RPC and native types. So this can be somewhat shorter too, just `UnmarshalCoinSelectionStrategy`.

guggero commented:
@mohamedawnallah can you please apply the fixup commits and rebase?

mohamedawnallah commented:
Okay, that makes sense I'm used to the monolith commits but I started to love the small commits with related changes that said I'm gonna update the commits individually so that it would be easier for review and let you know.

mohamedawnallah commented:
@guggero I've broken up the monolith commits into smaller commits please let me know if any feedback on the PR. Thanks 🙏

ellemouton commented on rpcserver.go:
```diff
@@ -2455,6 +2455,14 @@ func (r *rpcServer) BatchOpenChannel(ctx context.Context,
 		return nil, fmt.Errorf("specify at least one channel")
 	}
 
+	_, err := lnrpc.UnmarshallCoinSelectionStrategy(
+		in.CoinSelectionStrategy,
+		r.cfg.CoinSelectionStrategy,
+	)
+	if err != nil {
+		return nil, err
```
unused?

mohamedawnallah replied:
Yes forgot to remove it when used coin selection strategy as string in the beginning 👍

ellemouton commented on rpcserver.go:
```diff
@@ -1192,12 +1192,23 @@ func (r *rpcServer) EstimateFee(ctx context.Context,
+		in.CoinSelectionStrategy,
+		r.cfg.CoinSelectionStrategy,
+	)
+	if err != nil {
+		return nil, err
+	}
+
 	// We will ask the wallet to create a tx using this fee rate. We set
 	// dryRun=true to avoid inflating the change addresses in the db.
 	var tx *txauthor.AuthoredTx
 	wallet := r.server.cc.Wallet
 	err = wallet.WithCoinSelectLock(func() error {
-		tx, err = wallet.CreateSimpleTx(outputs, feePerKw, minConfs, true)
+		tx, err = wallet.CreateSimpleTx(
+			outputs, feePerKw, minConfs,
+			coinSelectionStrategy, true,
```
hmmm - i think this would have not compiled before this commit?

mohamedawnallah replied:
Yes, that's true. Thanks for this! I'm pretty new to the multiple small commits approach and I wasn't aware that each commit must be compiled successfully I was only looking for the last commit to be compiled successfully. I've just made the following dependency graph diagram. This will help me with which changes to make first in the commit history also I'll keep in mind to run `./scripts/check-each-commit.sh upstream/master` command next time.
<img width="751" alt="Screenshot 2024-03-12 at 9 58 42 PM" src="https://github.com/lightningnetwork/lnd/assets/69568555/9c166c1b-65ed-47b8-89bf-927523d0be2a">

ellemouton commented on lntest/mock/walletcontroller.go:
```diff
@@ -141,7 +141,8 @@ func (w *WalletController) ImportTaprootScript(waddrmgr.KeyScope,
 
 // SendOutputs currently returns dummy values.
 func (w *WalletController) SendOutputs([]*wire.TxOut,
-	chainfee.SatPerKWeight, int32, string) (*wire.MsgTx, error) {
+	chainfee.SatPerKWeight, int32, string,
```
i think this wont have compiled before this commit

ellemouton commented on lntest/mock/walletcontroller.go:
```diff
@@ -214,7 +214,7 @@ func (w *WalletController) ListLeasedOutputs() ([]*base.ListLeasedOutputResult,
 
 // FundPsbt currently does nothing.
 func (w *WalletController) FundPsbt(*psbt.Packet, int32, chainfee.SatPerKWeight,
-	string, *waddrmgr.KeyScope) (int32, error) {
```
dont think things calling this will compile

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -1091,6 +1091,18 @@ message LightningAddress {
+enum CoinSelectionStrategy {
+    // Use the coin selection strategy defined in the global configuration
+    // (lnd.conf).
+    STRATEGY_USE_GLOBAL_CONFIG = 0;
```
should we call it it 'default'? im open to being wrong here if other reviewers disagree 👍

ellemouton replied:
nvm - I agree with Oli that being more explicit here is good 👍

ellemouton commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,67 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+	switch strategy {
+	case CoinSelectionStrategy_STRATEGY_USE_GLOBAL_CONFIG:
+		strategyValue, _ := ParseCoinSelectionStrategy(globalStrategy)
+		coinSelectionStrategy, _ := UnmarshallCoinSelectionStrategy(
```
ok I see now that this wont actually loop but this is purely because we know that the string `globalStrategy` will be one of "random" or "largest". 

I think we can instead be a bit more strict about the structure here instead though:

[Here](https://github.com/lightningnetwork/lnd/blob/5cb4811e86ed1607c1c9671756d7276256e2ec44/config_builder.go#L590) is where we parse the "coin-selection-strategy" config option. We should only ever do that conversion from string to type once. Then, we can instead just return that wallet.CoinSelectionStrategy type here

mohamedawnallah commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,67 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+	*string, error) {
+
+	switch strategy {
+	case wallet.CoinSelectionLargest:
```
Thanks!

mohamedawnallah commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,67 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+// ParseCoinSelectionStrategyObject parses a wallet.CoinSelectionStrategy
+// object to its string counterpart.
+func ParseCoinSelectionStrategyObject(strategy wallet.CoinSelectionStrategy) (
+	*string, error) {
```
> why a pointer to a string? I think just return the string?

Yeah gonna use the empty string and found that approach followed in the codebase 👍

mohamedawnallah commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,67 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+// ParseCoinSelectionStrategy parses a coin selection strategy string
+// from the CLI to its lnrpc.CoinSelectionStrategy counterpart proto type.
+func ParseCoinSelectionStrategy(strategy string) (
+	*CoinSelectionStrategy, error) {
```
we return `nill` for `CoinSelectionStrategy` proto type incase if the strategy param doesn't match any of the following `global-config`, `largest`, or `random` if we don't use a pointer we get an error like `cannot use nil as CoinSelectionStrategy value in return statement`.

mohamedawnallah commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,67 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+
+	switch strategy {
+	case "global-config":
+		strategyVal := CoinSelectionStrategy_STRATEGY_USE_GLOBAL_CONFIG
```
Okay what should we return in case of no match ? `nil` gives an error since `CoinSelectionStrategy` is a proto type.
```
// ParseCoinSelectionStrategy parses a coin selection strategy string
// from the CLI to its lnrpc.CoinSelectionStrategy counterpart proto type.
func ParseCoinSelectionStrategy(strategy string) (
	CoinSelectionStrategy, error) {

	switch strategy {
	case "global-config":
		return CoinSelectionStrategy_STRATEGY_USE_GLOBAL_CONFIG, nil

	case "largest":
		return CoinSelectionStrategy_STRATEGY_LARGEST, nil

	case "random":
		return CoinSelectionStrategy_STRATEGY_RANDOM, nil

	default:
		return nil, fmt.Errorf("unknown coin selection strategy "+
			"%v", strategy)
	}
}
```Okay what should we return in case of no match ? `nil` gives an error since `CoinSelectionStrategy` is a proto type.
```
// ParseCoinSelectionStrategy parses a coin selection strategy string
// from the CLI to its lnrpc.CoinSelectionStrategy counterpart proto type.
func ParseCoinSelectionStrategy(strategy string) (
	CoinSelectionStrategy, error) {

	switch strategy {
	case "global-config":
		return CoinSelectionStrategy_STRATEGY_USE_GLOBAL_CONFIG, nil

	case "largest":
		return CoinSelectionStrategy_STRATEGY_LARGEST, nil

	case "random":
		return CoinSelectionStrategy_STRATEGY_RANDOM, nil

	default:
		return nil, fmt.Errorf("unknown coin selection strategy "+
			"%v", strategy)
	}
}
```

ellemouton replied:
you can just return `0` right? since it is an int under the hood and you are also returning an error and so the value wont be used

mohamedawnallah replied:
Yes makes sense. Thanks!

lightninglabs-deploy commented:
@guggero: review reminder
@ellemouton: review reminder
@mohamedawnallah, remember to re-request review from reviewers when ready

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -1091,6 +1091,18 @@ message LightningAddress {
+    // (lnd.conf).
+    STRATEGY_USE_GLOBAL_CONFIG = 0;
+
+    // Select the largest available coins first for spending.
```
suggest: "Select the largest available coins first during coin selection."

same below

ellemouton commented on config_builder.go:
```diff
@@ -587,18 +587,13 @@ func (d *DefaultWalletImpl) BuildWalletConfig(ctx context.Context,
-		return nil, nil, nil, fmt.Errorf("unknown coin selection "+
-			"strategy %v", d.cfg.CoinSelectionStrategy)
+	coinSelectionStrategy, err := lnrpc.MapCoinSelectionStrategy(
+		d.cfg.CoinSelectionStrategy,
+	)
+	if err != nil {
+		return nil, nil, nil, err
 	}
+	walletConfig.CoinSelectionStrategy = coinSelectionStrategy
```
I think we can leave this as is. We only want to do this conversion from config string to `wallet.CoinSelectionStrategy` _once_. It won't ever change. Then everywhere else where we need to know the global config, we can use this already parsed `wallet.CoinSelectionStrategy` value instead of re-parsing the string.

ellemouton replied:
sp perhaps all that is needed here is to make the `wallet.CoinSelectionStrategy` strategy more easily accessible from the rpcserver level

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -1145,6 +1162,10 @@ message SendManyRequest {
     // Whether unconfirmed outputs should be used as inputs for the transaction.
     bool spend_unconfirmed = 8;
+
+    // CoinSelectionStrategy is the strategy to use for selecting coins
```
This pattern of starting a comment with the name of the variable only applies to golang code btw. See the other examples here. 

So this can just start with "The coin selection strategy to use ...."

ellemouton replied:
same for other places here.

ellemouton commented on lnrpc/lightning.proto:
```diff
@@ -1105,6 +1117,11 @@ message EstimateFeeRequest {
     bool spend_unconfirmed = 4;
+
+    // CoinSelectionStrategy is the strategy to use for selecting coins
+    // to send bitcoin on-chain to multiple addresses for getting the fee
+    // estimates.
```
i think this is a liiiiitle confusing. As a user i dont think id understand what "multiple addresses" has to do with anything. I think you can just say that this is the coin selection strategy used during fee estimation. same for the others. Non blocking though - just my opinion.

ellemouton commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,64 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+
+// ParseCoinSelectionStrategy parses a coin selection strategy string
+// from the CLI to its lnrpc.CoinSelectionStrategy counterpart proto type.
+func ParseCoinSelectionStrategy(strategy string) (
+	CoinSelectionStrategy, error) {
+
+	switch strategy {
+	case "global-config":
+		return CoinSelectionStrategy_STRATEGY_USE_GLOBAL_CONFIG, nil
+
+	case "largest":
+		return CoinSelectionStrategy_STRATEGY_LARGEST, nil
+
+	case "random":
+		return CoinSelectionStrategy_STRATEGY_RANDOM, nil
+
+	default:
+		return 0, fmt.Errorf("unknown coin selection strategy "+
+			"%v", strategy)
+	}
```
this belongs in `cmd/lncli` since these strings are specific to lncli

ellemouton commented on lnrpc/marshall_utils.go:
```diff
@@ -179,3 +180,64 @@ func MarshalOutPoint(op *wire.OutPoint) *OutPoint {
+}
+
+// UnmarshallCoinSelectionStrategy converts a lnrpc.CoinSelectionStrategy proto
+// type to its wallet.CoinSelectionStrategy counterpart type.
```
perhaps mention what the globalStrategy param is used for

ellemouton commented on cmd/lncli/walletrpc_active.go:
```diff
@@ -848,6 +848,14 @@ var fundTemplatePsbtCommand = cli.Command{
+			Usage: "(optional) the strategy to use for selecting " +
+				"coins to fund the PSBT from a template. " +
+				"If provided it will override the globally " +
+				"configured strategy in lnd.conf",
```
should mention which options are available. 

Also, since this same flag is used in many places, might be worth extracting into a re-usable variable?

ellemouton commented on rpcserver.go:
```diff
@@ -1192,12 +1194,29 @@ func (r *rpcServer) EstimateFee(ctx context.Context,
 		return nil, err
 	}
 
+	strategy, err := lnrpc.MapCoinSelectionStrategy(
+		r.cfg.CoinSelectionStrategy,
+	)
+	if err != nil {
+		return nil, err
+	}
```
should not need to parse it again here

ellemouton commented on rpcserver.go:
```diff
@@ -1068,7 +1069,8 @@ func allowCORS(handler http.Handler, origins []string) http.Handler {
 // address to a specified output value to be sent to that address.
 func (r *rpcServer) sendCoinsOnChain(paymentMap map[string]int64,
 	feeRate chainfee.SatPerKWeight, minConfs int32,
-	label string) (*chainhash.Hash, error) {
+	label string,
+	strategy wallet.CoinSelectionStrategy) (*chainhash.Hash, error) {
```
nit: things can still fit on 3 lines instead of 4:

```
func (r *rpcServer) sendCoinsOnChain(paymentMap map[string]int64,
	feeRate chainfee.SatPerKWeight, minConfs int32, label string,
	strategy wallet.CoinSelectionStrategy) (*chainhash.Hash, error) {
```

ellemouton commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -980,7 +980,8 @@ func (b *BtcWallet) ImportTaprootScript(scope waddrmgr.KeyScope,
 // This is a part of the WalletController interface.
 func (b *BtcWallet) SendOutputs(outputs []*wire.TxOut,
 	feeRate chainfee.SatPerKWeight, minConfs int32,
-	label string) (*wire.MsgTx, error) {
+	label string,
+	strategy base.CoinSelectionStrategy) (*wire.MsgTx, error) {
```
```
func (b *BtcWallet) SendOutputs(outputs []*wire.TxOut,
	feeRate chainfee.SatPerKWeight, minConfs int32, label string,
	strategy base.CoinSelectionStrategy) (*wire.MsgTx, error) {
```

ellemouton commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -244,6 +244,9 @@
 * [Allow callers of ListSweeps to specify the start height](
   https://github.com/lightningnetwork/lnd/pull/7372).
 
+* Add coin selection strategy option to the following on-chain RPC calls
+  `EstimateFee`, `SendMany`, `SendCoins`, `BatchOpenChannel`, `SendOutputs`, and `FundPsbt`. [coin select: add coin selection strategy option to all on-chain RPCs](https://github.com/lightningnetwork/lnd/pull/8515).
```
notice that the other entries are wrapped at 80 chars :)

ellemouton reviewed (commented):
Looking good @mohamedawnallah ! 

Left some comments :)

mohamedawnallah commented:
Thank you very much, @ellemouton, for your feedback! The idea of converting the `CoinSelectionStrategy` string to `wallet.CoinSelectionStrategy` only once simplified the code significantly. Thanks for this. I've addressed all of your feedback :)

mohamedawnallah commented:
The latest change after asking for a review is related to the lining error solved by adding `//nolint:funlen` to `fundPsbtCoinSelect`:
```
lnrpc/walletrpc/walletkit_server.go:1474: Function 'fundPsbtCoinSelect' is too long (201 > 200) (funlen)
```

ellemouton commented on cmd/lncli/commands.go:
```diff
@@ -200,6 +200,14 @@ func newAddress(ctx *cli.Context) error {
+	Name: "coin_selection_strategy",
+	Usage: "(optional) the strategy to use for selecting " +
+		"coins. If provided, it will override the " +
+		"globally configured strategy in lnd.conf",
```
this should contain a list of the available options

ellemouton reviewed (commented):
tACK (tested via the FundPSBT flow) 🚀 

Great work! A few more suggestions 🙏

mohamedawnallah commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -845,13 +846,21 @@ func batchOpenChannel(ctx *cli.Context) error {
 	}
 
+	coinSelectionStrategy, err := parseCoinSelectionStrategy(
+		ctx.String("coin_selection_strategy"),
```
Yes, this makes total sense. Thanks!

ellemouton reviewed (approved):
LGTM! 🚀  Great work!

guggero commented on config.go:
```diff
@@ -378,6 +379,8 @@ type Config struct {
 
 	CoinSelectionStrategy string `long:"coin-selection-strategy" description:"The strategy to use for selecting coins for wallet transactions." choice:"largest" choice:"random"`
 
+	CoinSelectionStrategyObject wallet.CoinSelectionStrategy `description:"The coin selection strategy wallet object to use for selecting coins for wallet transactions." hidden:"true"`
```
This value is not required. You can simply access the value through `r.server.cc.Wallet.Cfg.CoinSelectionStrategy` in the RPC server.

guggero commented on cmd/lncli/main.go:
```diff
@@ -564,3 +564,31 @@ func networkParams(ctx *cli.Context) (*chaincfg.Params, error) {
 	}
 }
+
+// ParseCoinSelectionStrategy parses a coin selection strategy string
```
nit: Godoc should be lower case as well to match function name.

guggero commented on lnrpc/walletrpc/walletkit_server.go:
```diff
@@ -1174,10 +1181,18 @@ func (w *WalletKit) FundPsbt(_ context.Context,
+		coinSelectionStrategy base.CoinSelectionStrategy
 	)
 
+	coinSelectionStrategy, err = lnrpc.UnmarshallCoinSelectionStrategy(
```
nit: can just use `:=` operator here and then move `var feeSatPerKW chainfee.SatPerKWeight` to directly before the `switch` statement.

guggero commented on rpcserver.go:
```diff
@@ -1192,12 +1193,22 @@ func (r *rpcServer) EstimateFee(ctx context.Context,
-		tx, err = wallet.CreateSimpleTx(outputs, feePerKw, minConfs, true)
+		tx, err = wallet.CreateSimpleTx(
+			outputs, feePerKw, minConfs,
+			coinSelectionStrategy, true,
```
nit: looks like `coinSelectionStrategy` would fit on previous line.

guggero commented on docs/release-notes/release-notes-0.18.0.md:
```diff
@@ -280,6 +280,10 @@ bitcoin peers' feefilter values into account](https://github.com/lightningnetwor
   height](https://github.com/lightningnetwork/lnd/pull/7372).
 
+* Add coin selection strategy option to the following on-chain RPC calls
+  `EstimateFee`, `SendMany`, `SendCoins`, `BatchOpenChannel`, `SendOutputs`, and `FundPsbt`. [coin select: add coin selection strategy option
```
nit: this file should also break at 80 characters where possible.

guggero reviewed (commented):
Looks very good now! Just a couple of small requests, then we're good to go.

guggero commented on lnd.go:
```diff
@@ -443,6 +443,9 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
 		return mkErr("error creating wallet config: %v", err)
 	}
 
+	rpcServer.server.cc.Wallet.Cfg.CoinSelectionStrategy =
```
This is already set during the creation of the wallet.

mohamedawnallah replied:
Ah I see now, Thanks!

guggero reviewed (commented):
One last change required.

mohamedawnallah commented:
The latest push after asking for a review is just rebasing!

guggero reviewed (approved):
Nice, LGTM :tada:
